File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/binaryTreePaths_257.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

import java.util.ArrayList;
import java.util.List;

public class binaryTreePaths_257 {
    List<String> res;

    public List<String> binaryTreePaths(TreeNode root) {
        res = new ArrayList<>();
        helper(root, "");
        return res;
    }

    private void helper(TreeNode root, String s) {
        if (root == null) return;
        if (root.left == null && root.right == null) {
            s += root.val;
            res.add(s);
            return;
        }
        helper(root.left, s + root.val + "->");
        helper(root.right, s + root.val + "->");
    }
}

File: Clion/ZeroTrac/Tree/binaryTreePaths_257.cpp
//
// Created by TzashiNorpu on 10/31/2022.
//
#include "tree.h"

using namespace Tree;

vector<string> Solution::binaryTreePaths(TreeNode *root) {
  vector<string> res;
  function<void(TreeNode *, string)> helper = [&](TreeNode *node, string s) -> void {
      if (!node) return;
      if (!node->left && !node->right) {
        res.push_back(s + to_string(node->val));
        return;
      }
      helper(node->left, s + to_string(node->val) + "->");
      helper(node->right, s + to_string(node->val) + "->");
  };
  helper(root, "");
  return res;
}